paddle/fluid/operators/prune_gate_by_capacity_op.cu

```diff
@@ -0,0 +1,130 @@
+        platform::CudaAtomicAdd(expert_count_data + gate_idx_data[i], -1);
+    if (orig_cap <= 0) {
+      new_gate_idx_data[i] = -1;
+    } else {
+      new_gate_idx_data[i] = gate_idx_data[i];
+    }
+  }
+}
+
+template <typename DeviceContext, typename T1>
+class PruneGateByCapacityFunctor {
+ public:
+  PruneGateByCapacityFunctor(const framework::ExecutionContext& context,
+                             const framework::Tensor* gate_idx,
+                             framework::Tensor* expert_count_out,
+                             T1* new_gate_idx_data)
+      : context_(context),
+        gate_idx_(gate_idx),
+        expert_count_out_(expert_count_out),
+        new_gate_idx_data_(new_gate_idx_data) {}
+
+  template <typename T2>
+  void apply() {
+    auto batch_size = gate_idx_->numel();
+    auto* gate_idx_data = gate_idx_->data<T1>();
+
+    auto& dev_ctx = context_.template device_context<DeviceContext>();
+    auto* expert_count_out_data = expert_count_out_->data<T2>();
+
+    int blocks = NumBlocks(batch_size);
+    int threads = kNumCUDAThreads;
+
+    prune_gate_by_capacity_kernel<T1,
+                                  T2><<<blocks, threads, 0, dev_ctx.stream()>>>(
+        gate_idx_data, new_gate_idx_data_, expert_count_out_data, batch_size);
+
+#ifdef PADDLE_WITH_HIP
+    PADDLE_ENFORCE_CUDA_SUCCESS(hipStreamSynchronize(dev_ctx.stream()));
+#else
+    PADDLE_ENFORCE_CUDA_SUCCESS(cudaStreamSynchronize(dev_ctx.stream()));
+#endif
+  }
+
+ private:
+  const framework::ExecutionContext context_;
+  const framework::Tensor* gate_idx_;
+  framework::Tensor* expert_count_out_;
+  T1* new_gate_idx_data_;
+};
+
+template <typename Visitor>
+static void VisitDataType(framework::proto::VarType::Type type,
+                          Visitor visitor) {
+  if (type == framework::proto::VarType::INT32) {
+    visitor.template apply<int>();
+  } else if (type == framework::proto::VarType::INT64) {
+    visitor.template apply<int64_t>();
+  } else {
+    PADDLE_THROW(platform::errors::InvalidArgument(
+        "The recieved values gate_id type %s can not meet input requirements. "
+        "Because the given gate_id data type of operators must be int32 or "
+        "int64. Please input appropriate gate_id again! ",
+        framework::DataTypeToString(type)));
+  }
+}
+
+template <typename DeviceContext, typename T>
+class PruneGateByCapacityCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* gate_idx = context.Input<Tensor>("GateIdx");
+    auto* expert_count = context.Input<Tensor>("ExpertCount");
+    auto* expert_count_out = context.Output<Tensor>("ExpertCountOut");
+    auto* new_gate_idx = context.Output<Tensor>("NewGateIdx");
+    auto* new_gate_idx_data = new_gate_idx->mutable_data<T>(context.GetPlace());
+
+    framework::TensorCopy(*expert_count, context.GetPlace(), expert_count_out);
+    PruneGateByCapacityFunctor<DeviceContext, T> functor(
```

Comment:
这个地方这么调用的逻辑是什么?

Reply:
> 这个地方这么调用的逻辑是什么?

Because the input data has two input parameters, the data type may be different and needs to be handled for this situation.